### PR TITLE
Use the new async API for the metadata API and code generator

### PR DESCRIPTION
**Pull request by szeiger**

### CODE REVIEW

**Comment by cvogt** [comment]
@szeiger Why change the whole API of the ModelBuilder? What's the advantage of this?

**Comment by cvogt** [comment]
It matched the API of the code generator. The new ones seems to be unnecessarily different and break existing code. Why?

File: src/sphinx/upgrade.rst [hunk 1]
```diff
@@ -52,6 +52,22 @@ at the call site where the concrete type is known.
 ``Column[T]`` is still available as a deprecated alias for ``Rep[T]``. Due to the required implicit evidence, it
 cannot provide complete backwards compatibility in all cases.
 
+Closing Databases
+-----------------
+
+Since a ``Database`` instance can now have an associated connection pool and thread pool, it is important to call
+``close`` when you are done using it, so that these pools can be shut down properly. You should take care to do this
+wen you migrate to the new Action-based API. As long as you exclusively use the old API (and do not configure a
```
Comment (cvogt) [inline src/sphinx/upgrade.rst]: typo "wen"

**Comment by szeiger** [comment]
The old implementation relied on laziness and fetching data on demand from an implicit Session. This does not fit into the new asynchronous design. Blocking on all database results in order to keep the old design would not have worked well, either, because Sessions will go away. You can only run an Action directly on a Database, so there's no way to combine multiple calls on one Session outside of the Action monad.

I considered a fixpoint combinator for Actions to keep the lazy knot-tying but it wouldn't have solved this particular problem, either, plus it would have forced the code into the Future monad with the same consequences as running in the Action monad. (I may still add the fixpoint combinator though; it could be useful in other situations)

**Comment by cvogt** [comment]
I understand that part and I agree. I was merely talking about the names of the factory method in the virtual subclass pattern. def createPrimaryKeyBuilder instead of def PrimaryKey, etc. Why break the interface? The codegenerator uses the same naming convention and calling the factory like the class seems quite sensible to me, no need for pre and postfixing it with create and builder. That it is a builder is already clear from the package/type hierarchy context.

**Comment by szeiger** [comment]
The old design looks unnecessarily clever to me. I've never seen such naming conventions in other Scala code. Plus it only works with a single constructor argument, thus the nested classes. The new implementation has a cleaner separation between the different builders and more understandable (albeit longer) names.

**Comment by cvogt** [comment]
@szeiger I understand what you mean, but I am not sure I agree. It may be slightly unconventional (if that even exists in Scala), but the pattern beautifully matched the problem. And it was a very uniform design. And it exactly matched that of the code generator. Now they are different. I not saying the new design on its own is much worse, but I don't see the need for changing it and breaking the interface. Plus I like some things about the old design which are gone now, such as the nesting, which made it easy to override behavior for individual columns, etc. without making a case distinction in the overridden method, which you would have to do now, because they are all top-level if I am not mistaken.

**Comment by szeiger** [comment]
WDYM? Previously you had to create both a TableBuilder and a ColumnBuilder to customize the ColumnBuilder, now you can do it directly. If you want to customize only the columns of specific tables, you need to make the distinction at some place, whether that is when you create the TableBuilder or the ColumnBuilder doesn't really matter.

**Pull request merged**